### PR TITLE
515 Compatibility

### PR DESCRIPTION
**Pull request by AnturK**
Makes the code compatible with 515.1594+

Few simple changes and one very painful one.
Let's start with the easy:
* puts call behind `LIBCALL` define, so call_ext is properly used in 515
* Adds `NAMEOF_STATIC(_,X)` macro for nameof in static definitions since src is now invalid there.
* Fixes tgui and devserver. From 515 onward the tmp3333{procid} cache directory is not appened to base path in browser controls so we don't check for it in base js and put the dev server dummy window file in actual directory not the byond root.
* Renames the few things that had /final/ in typepath to ultimate since final is a new keyword

And the very painful change:
`.proc/whatever` format is no longer valid, so we're replacing it with new nameof() function. All this wrapped in three new macros.
`PROC_REF(X)`,`TYPE_PROC_REF(TYPE,X)`,`GLOBAL_PROC_REF(X)`. Global is not actually necessary but if we get nameof that does not allow globals it would be nice validation.
This is pretty unwieldy but there's no real alternative.
If you notice anything weird in the commits let me know because majority was done with regex replace.

@tgstation/commit-access Since the .proc/stuff is pretty big change.

### CODE REVIEW

**Comment by san7890** [comment]
Should we bump `dependencies.sh` now or should that be further down the line?

**Review by MrMelbert — CHANGES_REQUESTED** [review]
Forgot to repath `/datum/traitor_objective/final/supermatter_cascade/can_generate_objective(generating_for, list/possible_duplicates)`. Causing tests to fail

**Comment by MrStonedOne** [comment]
We will likely need a 515 specific compile check (either a compile maps or a windows build or at least some thing to confirm everything compiles) or new prs will introduce 515 incompatible code until we switch and find out.

**Comment by AffectedArc07** [comment]
> Should we bump `dependencies.sh` now or should that be further down the line?

We put in a system that supports both versions in CI (https://github.com/ParadiseSS13/Paradise/pull/19571), however I wouldnt bump the main master versions, largely because servers will (presumably) still be on 514, and it could easily turn into an issue where the prod servers cant build.

Also SDMM wont be updateable yet if you bump the required compile version in `_compile_options.dm`

**Comment by MrStonedOne** [comment]
> Should we bump `dependencies.sh` now or should that be further down the line?

That should match which on the game servers unless you want to deal with drifts and conflicts and failed updates.

but we do need a way to confirm new code is still 515 safe. and after the game servers move to 515, we will need the same thing for 514 until we move to upping min compile version which we try not to do until the client side bugs are fixed since devs generally have to use the same version for both client and server unless they want to upend their workflow.

**Comment by AnturK** [comment]
I think ci matrix would be good, probably locked behind some secret so we can turn it off/on on demand without commits. One problem here is the .proc/stuff which will probably be a majority of these incompatibilities actually passes compile but behavior differs at runtime.

**Comment by MrStonedOne** [comment]
I think ci matrix would be good, probably locked behind some secret so we can turn it off/on on demand without commits. 

fuck you, no.

any collaborator must be able to do this in a pr.

don't fucking try to lock that away into the ivory tower.

**Comment by AnturK** [comment]
> > I think ci matrix would be good, probably locked behind some secret so we can turn it off/on on demand without commits.
> 
> fuck you, no.
> 
> any collaborator must be able to do this in a pr.
> 
> don't fucking try to lock that away into the ivory tower.

I have no idea what you're trying to say. I'm talking about automated check.

**Comment by Mothblocks** [comment]
I think he is talking about how secrets don't apply to PRs since it uses the secrets on your fork

**Comment by Mothblocks** [comment]
We only need one map for a 515 test, and we do need to test more than just compile (like runtime call_ext failures if it can't statically infer the warning). I recommend just having a Runtime Station on 515 workflow.

**Comment by MrStonedOne** [comment]
no, i mean how if its a secret, then one would have to ask maintainers to do it.

if its a file, anybody can change it and pr the change with the new versions to check without having to ask maintainers in discord to do it.

I want more coder/collaborator/non-maintainer control over the repo, having the .github folder enables that, anybody can change the settings in there and pr the change for approval.

Moving something from the github actions folder to the secrets panel locks people out of being able to pr changes to it.

this includes a config secret like 'run alt byond version' i shouldn't have to ask maintainers/headcoders to change that over discord, i should be able to ask them to change it by opening a pr that changes it.

**Comment by Mothblocks** [comment]
If I'm understanding right, we are talking about not wanting to just run a 515 test on every PR? Either because they might not have the PR merged, or because we'd hit this problem again when 516 etc come out?

I think a good way to trip that is something like `#test_versions: 515.1234` in a PR body, and having an action read that?

File: code/modules/antagonists/heretic/knowledge/sacrifice_knowledge/sacrifice_knowledge.dm
Comment (msgerbs) [inline code/modules/antagonists/heretic/knowledge/sacrifice_knowledge/sacrifice_knowledge.dm]:
```suggestion
 * This proc is called from [proc/sacrifice_process] after the heretic successfully sacrifices [sac_target].
```

File: code/modules/antagonists/heretic/knowledge/sacrifice_knowledge/sacrifice_knowledge.dm
Comment (msgerbs) [inline code/modules/antagonists/heretic/knowledge/sacrifice_knowledge/sacrifice_knowledge.dm]:
```suggestion
 * This proc is called from [proc/begin_sacrifice] after the [sac_target] falls asleep, shortly after the sacrifice occurs.
```

File: code/modules/antagonists/heretic/knowledge/sacrifice_knowledge/sacrifice_knowledge.dm
Comment (msgerbs) [inline code/modules/antagonists/heretic/knowledge/sacrifice_knowledge/sacrifice_knowledge.dm]:
```suggestion
 * This proc is called from [proc/after_target_sleeps] when the [sac_target] should be waking up.
```

File: code/modules/antagonists/heretic/knowledge/sacrifice_knowledge/sacrifice_knowledge.dm
Comment (msgerbs) [inline code/modules/antagonists/heretic/knowledge/sacrifice_knowledge/sacrifice_knowledge.dm]:
```suggestion
 * This proc is called from [proc/after_target_wakes] after the helgrasp runs out in the [sac_target].
```

File: code/modules/antagonists/heretic/knowledge/sacrifice_knowledge/sacrifice_knowledge.dm
Comment (msgerbs) [inline code/modules/antagonists/heretic/knowledge/sacrifice_knowledge/sacrifice_knowledge.dm]:
```suggestion
 * This proc is called from [proc/begin_sacrifice] if the target survived the shadow realm, or [COMSIG_LIVING_DEATH] if they don't.
```

File: code/modules/antagonists/heretic/knowledge/sacrifice_knowledge/sacrifice_knowledge.dm
Comment (msgerbs) [inline code/modules/antagonists/heretic/knowledge/sacrifice_knowledge/sacrifice_knowledge.dm]:
```suggestion
 * This proc is called from [proc/return_target] if the [sac_target] survives the shadow realm.
```

File: code/modules/antagonists/heretic/knowledge/sacrifice_knowledge/sacrifice_knowledge.dm
Comment (msgerbs) [inline code/modules/antagonists/heretic/knowledge/sacrifice_knowledge/sacrifice_knowledge.dm]:
```suggestion
 * This proc is called from [proc/return_target] if the target dies in the shadow realm.
```

**Comment by MrStonedOne** [comment]
no. you need to read the message i was replying to originally to understand my complaint.

File: code/modules/antagonists/nukeop/equipment/nuclear_bomb/_nuclear_bomb.dm
Comment (msgerbs) [inline code/modules/antagonists/nukeop/equipment/nuclear_bomb/_nuclear_bomb.dm]:
```suggestion
 * [proc/explode] -> [proc/actually_explode] -> [proc/really_actually_explode]
```

**Comment by MrStonedOne** [comment]
i don't want it to be able to turn on and off without commits, because that also means people can't open a pr to turn it on for the whole repo

**Comment by MrStonedOne** [comment]
i want more ability for people to pr changes to this repo's settings.

this means not locking non-secrets behind secrets, and keeping them rightfully within the `.github` folder.

**Comment by Mothblocks** [comment]
I did read it but it wasn't entirely clear to me, yes I agree we shouldn't want a secret for anything relating to this

File: code/modules/client/client_procs.dm
Comment (msgerbs) [inline code/modules/client/client_procs.dm:133]:
```suggestion
///dumb workaround because byond doesnt seem to recognize the .proc/Topic() typepath for /datum/proc/Topic() from the client Topic,
```

Also is the thing this is a workaround for entirely different in BYOND 515?

File: code/__HELPERS/areas.dm
Comment (Mothblocks) [inline code/__HELPERS/areas.dm:44]: Your find and replace hit way too much documentation

Comment (MrStonedOne) [reply code/__HELPERS/areas.dm:44]: What i never understand is why people do this blindly so often.

every time i have regex replaced a thing, i manually hit replace next like 100 to 500 times before i hit replace all.

is that really so fucking hard.

Comment (AnturK) [reply code/__HELPERS/areas.dm:44]: Yeah, i'll try to clean these up in a moment

Comment (MrStonedOne) [reply code/__HELPERS/areas.dm:44]: also we clearly need a spacemandmm system for regex code replacements. being able to pass in shit like --skip-comments or --only-comments or --proc-bodies-only would help so so so so much.

Comment (AnturK) [reply code/__HELPERS/areas.dm:44]: yeah, would be nice. It's pain to do by hand.

File: code/modules/mining/lavaland/tendril_loot.dm
Comment (msgerbs) [inline code/modules/mining/lavaland/tendril_loot.dm:930]:
```suggestion
	)
```

File: code/modules/mob/living/brain/brain_item.dm
Comment (msgerbs) [inline code/modules/mob/living/brain/brain_item.dm:439]:
```suggestion
/// DOES NOT handle setting up the trauma, that's done by [proc/brain_gain_trauma]!
```

File: code/modules/reagents/chemistry/reagents/other_reagents.dm
Comment (msgerbs) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm:1435]:
```suggestion
//For colouring in [/proc/mix_color_from_reagents]
```

File: code/modules/spells/spell.dm
Comment (msgerbs) [inline code/modules/spells/spell.dm]:
```suggestion
 * - SPELL_NO_FEEDBACK will prevent the spell from calling [proc/spell_feedback] on cast. (invocation, sounds)
```

File: code/modules/spells/spell_types/jaunt/bloodcrawl.dm
Comment (msgerbs) [inline code/modules/spells/spell_types/jaunt/bloodcrawl.dm]:
```suggestion
 * and calling [proc/on_victim_consumed] if successful.
```

File: code/modules/surgery/bodyparts/_bodyparts.dm
Comment (msgerbs) [inline code/modules/surgery/bodyparts/_bodyparts.dm:40]:
```suggestion
	///The actual color a limb is drawn as, set by [/proc/update_limb]
```

**Review by msgerbs — COMMENTED** [review]
I'm sure most of these were fixed while I was going through them all by hand but oh well. I made a few of these into autodoc links since I was already 90% of the way there,

**Comment by san7890** [comment]
@AnturK To save you the few minutes, I made a PR to fix the merge conflict here: https://github.com/AnturK/tgstation/pull/10

**Comment by MrStonedOne** [comment]
you can't just fix merge conflicts with a merge! it doesn't work like that!

**Comment by Mothblocks** [comment]
@AnturK just for posterity do you think you could briefly explain the 515 bug you're facing

**Comment by AnturK** [comment]
```
    //var/list/stuff = range(1,src)
    for(var/atom/close_atom as anything in range(1,src)) // Replacing with ... in stuff) makes this work
        log_game("[ritual.type] atom in range loop [close_atom]") // never hit
        if(!ismovable(close_atom))
            continue
        if(close_atom.invisibility)
            continue
        if(close_atom == user)
            continue

        atoms_in_range += close_atom
```
In short the heretic bug goes down to for over range(1,src) returning nothing, exact same loop works when range is placed in intermediate variable. There were some optimizations to this in 515 so that's likely cause. Unfortunately this is reproducible in full repo but not in simple test project.

**Comment by Mothblocks** [comment]
Is it reasonable for us to, only for the sake of moving on and seeing what else is potentially broken, `#define` range away with our own special proc? Not to be merged into master, but just to see if there's any other bugs we find that will have to be fixed anyway?

File: .github/guides/STANDARDS.md
Comment (AffectedArc07) [inline .github/guides/STANDARDS.md:120]: Maybe write a bit on which ref you should use?

Comment (AffectedArc07) [reply .github/guides/STANDARDS.md:120]: Wait this isnt my repo

Comment (Mothblocks) [reply .github/guides/STANDARDS.md:120]: This should go into RegisterSignal documentation at the very least

**Comment by Mothblocks** [comment]
Lummox says the issue is resolved. Do we know when we can expect a new release?

File: .github/guides/STANDARDS.md
Comment (AnturK) [reply .github/guides/STANDARDS.md:164]: Added some basic examples/instructions

File: .github/guides/STANDARDS.md
Comment (Mothblocks) [reply .github/guides/STANDARDS.md:102]: The docs for RegisterSignal I mean

**Comment by Mothblocks** [comment]
told u so :(

**Comment by Mothblocks** [comment]
The worrying part is it can't find any references to the hard deleted stuff in the test

https://github.com/tgstation/tgstation/actions/runs/3444022611/jobs/5746182839

CC @LemonInTheDark

File: code/modules/unit_tests/metabolizing.dm
Comment (Mothblocks) [inline code/modules/unit_tests/metabolizing.dm:7]: Note to self: We should make a separate test for this then

**Comment by LemonInTheDark** [comment]
> The worrying part is it can't find any references to the hard deleted stuff in the test

I've been keeping an eye on this while anturk has been ranting and I genuinely have no idea.
He's bullying lummy so I wanna see what comes of that.

**Comment by Mothblocks** [comment]
Good luck, hoping we can get this through and lift freeze

File: tools/ci/check_grep.sh
Comment (san7890) [inline tools/ci/check_grep.sh:308]:
```suggestion
    echo "${RED}ERROR: Outdated proc reference use detected in code, please use proc reference helpers.${NC}"
```

Consistency with the rest of check_grep.sh (colors provide good contrast)

**Comment by Mothblocks** [comment]
As an update, we are not strictly blocked on tests but we are blocked on a backwards incompatible issue with savefiles to where we would not be able to switch between 515 and 514 without breaking people

File: code/__byond_version_compat.dm
Comment (MrStonedOne) [reply code/__byond_version_compat.dm:20]: who cares, if it works like we want we get the feature and if it doesn't its dead code until somebody nags lummox.

**Comment by MrStonedOne** [comment]
> As an update, we are not strictly blocked on tests but we are blocked on a backwards incompatible issue with savefiles to where we would not be able to switch between 515 and 514 without breaking people

are you sure?

because there is code to make it not do that in the pr.

if you have an argument for why that won't work you have to state it in the pr.

**Comment by Mothblocks** [comment]
I'm just reporting what AnturK told me because it wasn't posted here @MrStonedOne

File: code/__byond_version_compat.dm
Comment (AnturK) [reply code/__byond_version_compat.dm:20]: Well, it's not like we change min compiler define without testing, i think it's fine. Worst case you can just change it on the breaking version.

**Comment by Mothblocks** [comment]
We're going to be merging this but with a preprocessor check to block usage of 515, which AnturK says is not backwards compatible right now, just so that we can lift the freeze

File: code/__byond_version_compat.dm
Comment (ZeWaka) [reply code/__byond_version_compat.dm:20]: > also it helps to read the documentation. Download 515 in zip form. opem up dreammaker, and hit f1 and browse the vars for savefile.

the docs on the byond website are the 515 ones, it's annoying yet handy

**Pull request merged**
